### ISSUE TITLE
Distributed training

### ISSUE THREAD

**Issue by bhack** [tensorflow/tfjs-core#35]
Is there a plan to experiment a [distributed training](https://arxiv.org/abs/1702.01846)?

**Comment by kaihuchen** [comment]
My opinion is that since deeplearnjs aims to provide ML functions at a higher level similar to that of TensorFlow, it is not meant for supporting the kind of distributed training at the **weight level** as mentioned in the article.

However, I do believe that deeplearnjs is a very good candidate for carrying out a kind of massive-scale crowd-driven **ensemble leaning**.

Let's say you have a massive learning task to solve, so you appeal to the public to help, by asking them to contribute their GPUs through web browser. The general idea of the ensemble learning is as follows (and many variations are certainly possible):

- The training dataset is divided into many small chunks, and many different types of candidate models are created (ensemble learning works better with diversity) for the task. 
- Each contributing client gets a small training dataset and a candidate model to be trained in their browser. When completed the trained model Ti is then uploaded to the server.
- Each trained model Ti is tested against a test dataset, and receives a confidence score Si regarding how good it is with the task. The combined model set {Ti} is also filtered to remove ineffective ones.
- The combined model set {Ti} and the confidence scores {Si} are then used jointly for making predictions, which likely will yield more precise prediction than any single model

Such an ensemble learning approach has many advantages:

- Can be massively scalable
- When compared with the monolithic model approach, it is perhaps competitive in performance (my guess here, need proof), and faster (assuming that there are many willing GPU contributors).
- Can be capable of incremental learning. I.e., we can retire out-of-date models or add newly trained models (say, when new classes show up, old classes got merged, when distribution changes, etc.) at anytime easily, in order to keep the combined model set {Ti} robust and accurate.

**Comment by nsthorat** [comment]
+1 to @kaihuchen's comments!

As of now, we don't have plans to do this, but we're accepting pull requests ;)

**Comment by dsmilkov** [comment]
Re-opening this to keep the discussion flowing.

I think an experiment for crowd-driven ensemble learning would be very very cool. We don't have any plans to do this, but if anyone is interested, we'll be happy to provide guidance and support.

**Comment by kaihuchen** [comment]
@dsmilkov   Much appreciated for your support! To get it going I will proceed to spell out more details regarding the approach, other benefits, as well as dependencies. Hopefully we can also attract more people to contribute to the effort.

**Comment by bhack** [comment]
Another interesting resource https://arxiv.org/abs/1703.05422

**Comment by bhack** [comment]
And the follow-up work [Developing a Volunteer Computing Project to Evolve Convolutional Neural Networks and Their Hyperparameters](http://tdesell.cs.und.edu/papers/2017_escience_exact.pdf)

**Comment by bhack** [comment]
https://github.com/travisdesell/exact /cc @travisdesell

**Comment by kaihuchen** [comment]
@bhack  Thanks for sharing. I could see that deeplearnjs could be quite useful for such an evolutional approach.

**Comment by kaihuchen** [comment]
There are many ways to use deeplearnjs for the ensemble learning approach. Below is my attempt to find a workable path.

One way to do ensemble learning is the **generalist approach**. For example, assuming that we wish to train a system to recognize every type of animal in the world, so we divide up the dataset to many smaller chunks, and train many models in an ensemble for the task. Each model in the ensemble is meant to recognize everything, but perhaps by using a different type of model (say, Alexnet vs Inception vs Squeezenet, etc). Such a generalist recognizer is broad and shallow, i.e., it can recognize many things in general terms (e.g., recognizing the image is a bird), but it does not learn something that is narrow and specific (e.g., recognizing a given image as a Black Footed Albatross, out of 22 different types of albatrosses).

Another approach that I thought is more promising is the **specialist approach**, where each model in an ensemble {Ti} are trained to become an expert in recognizing very specific things (e.g., distinguishing among the 22 types of albatrosses) and nothing else. 

The benefits of going with the **specialist approach** are as follows:

- Obviously there are cases where we want the specifics, and not just the more general label.
- **It can be used to achieve the effect of [one-shot learning](https://en.wikipedia.org/wiki/One-shot_learning)**.  One-shot learning depends on the existence of prior knowledge, and the *trained experts* in the specialist approach are in fact our background knowledge for achieving one-shot learning. 
For example, let's say we have trained the system to recognize all kinds of animals, with some models being the *albatross expert*, some being the *bird expert*, some being the *dog expert*, some being the *insect expert*, etc. Now if the system is being shown the image of a previously unseen animal (say, the mythical *Qilin*), then system can receive a set of activations from the set of trained experts in the ensemble. For example, overall the system might recognize a Qilin as 40% horse, 30% deer, 25% lizard, etc. So you can say that this set of expert activations is now our coordinate for the prototypical Qilin, it is then possible to recognize another Qilin image, even if the system has seen it only once.
- **It enables a simple form of [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning)**. In the Qilin example above, let's say that eventually we obtained some more training samples for Qilin, so we want to train a new model for recognizing the Qilin class. Here we don't have to start this new model from scratch, but rather we can make a copy of one of the top *contributing expert* (which was the **horse** expert above), or perhaps by combining multiple top contributing expert models into one, and train further with it.
- **Enables customization**. For example, a client is able to train a model to recognize specifically the members of his family, and nothing else.

**Comment by kaihuchen** [comment]
**Or to put it more succinctly**: I believe that using deeplearnjs with the *specialist ensemble* approach will give us a very good starting point for building a crowd-contributed world-scale system for image recognition of just about anything (subject to the limitations of the underlying recognition model).

**Comment by nsthorat** [comment]
@kaihuchen I love the vision. Batching is coming soon, but in the meantime we can provide support if you want to start prototyping. We can also sync over a video conference call if you want to chat in more detail.

**Comment by kaihuchen** [comment]
@nsthorat  Very excited to have your support for this! I do believe that deeplearnjs can be used for many novel and interesting purposes, and I'd love to work closely with your team to make it happen. Give me a few days to work out a rough plan, and I will reach out to you to discuss it further.

**Comment by nsthorat** [comment]
I still think this is an incredible idea - feel free to start a conversation in our TensorFlow.js community mailing list if you want to continue this discussion: https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs

**Comment by travisdesell** [comment]
@kaihuchen @bhack sorry I must have missed this in my inbox! I would be interested in seeing what parts of EXACT I could port over to this.  In particular I've been thinking about further separating out the implementation which evolves the structure of the networks and what actually does the training.

That being said, EXACT is implemented in C++ right now (so that I can use it on my volunteer computing project).  I'm not sure how easy it would be to get this to integrate with javascript.
